floodlight/io/datasets.py

```diff
@@ -265,3 +260,373 @@ def get_pitch(self) -> Pitch:
+        filepath_events = os.path.join(
+            self._events_data_dir,
+            str(mID) + self._STATSBOMB_FILE_EXT,
+        )
+        filepath_threesixty = os.path.join(
+            self._threesixty_data_dir,
+            str(mID) + self._STATSBOMB_FILE_EXT,
+        )
+
+        # check if events need to be downloaded
+        if not os.path.exists(filepath_events):
+            events_host_url = (
+                f"{self._STATSBOMB_SCHEMA}://"
+                f"{self._STATSBOMB_BASE_URL}/"
+                f"{self._STATSBOMB_EVENTS_FOLDERNAME}/"
+                f"{str(mID)}"
+                f"{self._STATSBOMB_FILE_EXT}"
+            )
+            with open(filepath_events, "wb") as binary_file:
+                binary_file.write(download_from_url(events_host_url))
+
+        # check if StatsBomb360 data is available and needs to be downloaded
+        if not os.path.exists(filepath_threesixty):
+            threesixty_host_url = (
+                f"{self._STATSBOMB_SCHEMA}://"
+                f"{self._STATSBOMB_BASE_URL}/"
+                f"{self._STATSBOMB_THREESIXTY_FOLDERNAME}/"
+                f"{str(mID)}"
+                f"{self._STATSBOMB_FILE_EXT}"
+            )
+            try:
+                data = download_from_url(threesixty_host_url)
+                with open(filepath_threesixty, "wb") as binary_file:
+                    binary_file.write(data)
+            except HTTPError:
+                filepath_threesixty = None
+
+        # read events from file
+        (home_ht1, home_ht2, away_ht1, away_ht2,) = read_open_statsbomb_event_data_json(
+            filepath_events, filepath_matches, filepath_threesixty
+        )
+        event_objects = (home_ht1, home_ht2, away_ht1, away_ht2)
+
+        return event_objects
+
+    @staticmethod
+    def get_pitch() -> Pitch:
+        """Returns a Pitch object corresponding to the StatsBomb Dataset."""
+        return Pitch.from_template("statsbomb", sport="football")
+
+    def _read_competition_season_and_match_links_from_files(self):
+        """Creates the dictionaries containing data links between competition, season,
+        and matches to the respective cIDs, sIDs, and mIDs for all available matches and
+        every competition and season in the StatsBomb dataset.
+        """
+        # updates on competition level
+        competition_info = pd.read_json(self.filepath_competitions)
+        cIDs = competition_info["competition_id"].unique()
+        competitions = competition_info["competition_name"].unique()
+        self.links_competition_to_cID.update(
+            {competition: cIDs[i] for i, competition in enumerate(competitions)}
+        )
+        self.links_season_to_sID.update(
+            {competition: {} for competition in competitions}
+        )
+        self.links_match_to_mID.update(
+            {competition: {} for competition in competitions}
+        )
+
+        # loop
+        for _, single_season in competition_info.iterrows():
```

Comment:
I believe the bottleneck here is not the File I/O but rather the download of the info files in the first place (matches/2/44.json etc.). 

However, I still think it is useful to download these information in one go with the competitions.json rather than on demand. 

Coming back to your point I have now divided the _read_links() function into two parts: 1. update_competition_links called upon instantiation and 2. update_season_match_links(competition_name) only called on demand during the get() function. 

Let me know what you think of that solution and whether you agree that downloading match info on demand is better.